styleganv2/networks/shared/custom_layers.py

```diff
@@ -0,0 +1,295 @@
+        # random normal distribution
+        torch.nn.init.normal_(self.weight)
+        if bias:
+            torch.nn.init.zeros_(self.bias)
+
+        # define the scale for the weights:
+        fan_in = self.in_features
+        self.scale = np.sqrt(2) / np.sqrt(fan_in)
+
+    def forward(self, x: Tensor) -> Tensor:
+        return torch.nn.functional.linear(x, self.weight * self.scale, self.bias)
+
+
+class AdaIN(Module):
+    """
+    AdaIN layer
+    Args:
+        mod_vector_channels: number of channels in the modulating vector
+        num_channels: number of channels in the input_volume
+    """
+
+    def __init__(self, mod_vector_channels: int, num_channels: int) -> None:
+        super().__init__()
+        self.mod_vector_channels = mod_vector_channels
+        self.num_channels = num_channels
+        self.mu_transformer = EqualizedLinear(mod_vector_channels, num_channels)
+        self.sigma_transformer = EqualizedLinear(mod_vector_channels, num_channels)
+
+    def forward(self, x: Tensor, mod_vector: Tensor, alpha=1e-8) -> Tensor:
+        """
+        applies the AdaIN transform on the input
+        Args:
+            x: input volume [B x c x h x w]
+            mod_vector: vector for obtaining the AdaIN [B x d]
+            alpha: a very small value for numerical stability
+        Returns: Adaptively instance normalized input
+        """
+        # instance normalize x:
+        x_in = (x - x.mean(dim=(-1, -2), keepdim=True)) / (
+            x.std(dim=(-1, -2), keepdim=True) + alpha
+        )
+
+        # obtain the new mus and sigmas
+        mus = self.mu_transformer(mod_vector)
+        sigmas = 1 + self.sigma_transformer(mod_vector)
+
+        # reshape the mus and sigmas properly
+        batch_size, num_channels = mus.shape
+        mus, sigmas = (
+            mus.reshape((batch_size, num_channels, 1, 1)),
+            sigmas.reshape((batch_size, num_channels, 1, 1)),
+        )
+
+        # apply the mus and sigmas to the instance normalized x
+        return (x_in * sigmas) + mus
+
+
+class NoOp(Module):
+    """ Useful for code readability """
+
+    @staticmethod
+    def forward(x: Any) -> Any:
+        return x
+
+
+class MinibatchStdDev(Module):
+    """
+    Minibatch standard deviation layer for the discriminator
+    Args:
+        group_size: Size of each group into which the batch is split
+        num_new_features: number of additional feature maps added
+    """
+
+    def __init__(self, group_size: int = 4, num_new_features: int = 1) -> None:
+        """
+
+        Args:
+            group_size:
+            num_new_features:
+        """
+        super(MinibatchStdDev, self).__init__()
+        self.group_size = group_size
+        self.num_new_features = num_new_features
+
+    def extra_repr(self) -> str:
+        return f"group_size={self.group_size}, num_new_features={self.num_new_features}"
+
+    def forward(self, x: Tensor, alpha: float = 1e-8) -> Tensor:
```

Comment:
OMG wow! This is the slowest implementation of MinBatchStdDev I have ever seen. It's going to take magnitudes of more time compared to original TF implementation. Refer to the original code to make this better.